### ISSUE TITLE
Android support

### ISSUE THREAD

**Issue by Rudloff** [desktop/dugite#146]
Hello,

When trying to install dugite on Android, I get this error:
```
yarn add v1.3.2
info No lockfile found.
[1/4] Resolving packages...
[2/4] Fetching packages...
[3/4] Linking dependencies...
[4/4] Building fresh packages...
[1/1] ⠠ dugite
[-/1] ⠐ waiting...
[-/1] ⠐ waiting...
[-/1] ⠐ waiting...
error /data/data/com.termux/files/home/node_modules/dugite: Command failed.
Exit code: 1
Command: node ./script/download-git.js
Arguments: 
Directory: /data/data/com.termux/files/home/node_modules/dugite
Output:
/data/data/com.termux/files/home/node_modules/dugite/script/config.js:32
  const index = pathName.lastIndexOf('/')
                         ^

TypeError: Cannot read property 'lastIndexOf' of null
    at getConfig (/data/data/com.termux/files/home/node_modules/dugite/script/config.js:32:26)
    at Object.<anonymous> (/data/data/com.termux/files/home/node_modules/dugite/script/download-git.js:11:35)
    at Module._compile (module.js:635:30)
    at Object.Module._extensions..js (module.js:646:10)
    at Module.load (module.js:554:32)
    at tryModuleLoad (module.js:497:12)
    at Function.Module._load (module.js:489:3)
```

This is because `process.platform == 'android'`.

I don't expect you to provide an ARM Android binary but I understand we can use another Git binary with `GIT_EXEC_PATH`, so I guess the install script could just ignore the download (maybe with a warning) when on Android.

**Comment by shiftkey** [comment]
This is available in `1.50.0` on NPM - try it out and see how it feels!

**Comment by Rudloff** [comment]
Thanks a lot for merging this quickly but I still get an error when tring to use `exec()`:
```
Promise {
  <rejected> Error: Git not supported on platform: android
    at resolveGitBinary (/data/data/com.termux/files/home/.config/yarn/global/node_modules/dugite/build/lib/git-environment.js:35:11)
    at Object.setupEnvironment (/data/data/com.termux/files/home/.config/yarn/global/node_modules/dugite/build/lib/git-environment.js:65:25)
    at /data/data/com.termux/files/home/.config/yarn/global/node_modules/dugite/build/lib/git-process.js:53:60
    at new Promise (<anonymous>)
    at Function.exec (/data/data/com.termux/files/home/.config/yarn/global/node_modules/dugite/build/lib/git-process.js:48:16)
    at repl:1:19
    at ContextifyScript.Script.runInThisContext (vm.js:50:33)
    at REPLServer.defaultEval (repl.js:240:29)
    at bound (domain.js:301:14)
    at REPLServer.runBound [as eval] (domain.js:314:12),
  domain: 
   Domain {
     domain: null,
     _events: { error: [Function: debugDomainError] },
     _eventsCount: 1,
     _maxListeners: undefined,
     members: [] } }
```

`LOCAL_GIT_DIRECTORY` and `GIT_EXEC_PATH` are both defined:
```js
{ ANDROID_DATA: '/data',
  ANDROID_ROOT: '/system',
  EXTERNAL_STORAGE: '/sdcard',
  HOME: '/data/data/com.termux/files/home',
  LANG: 'en_US.UTF-8',
  LD_LIBRARY_PATH: '/data/data/com.termux/files/usr/lib',
  LD_PRELOAD: '/data/data/com.termux/files/usr/lib/libtermux-exec.so',
  LOGNAME: 'u0_a118',
  PATH: '/data/data/com.termux/files/usr/bin:/data/data/com.termux/files/usr/bin/applets',
  PREFIX: '/data/data/com.termux/files/usr',
  PWD: '/data/data/com.termux/files/home/.config/yarn/global',
  SHELL: '/data/data/com.termux/files/usr/bin/fish',
  SHLVL: '1',
  SSH_CLIENT: '127.0.0.1 36633 8022',
  SSH_CONNECTION: '127.0.0.1 36633 127.0.0.1 8022',
  SSH_TTY: '/dev/pts/1',
  TERM: 'xterm-256color',
  TMPDIR: '/data/data/com.termux/files/usr/tmp',
  USER: 'u0_a118',
  LOCAL_GIT_DIRECTORY: '/data/data/com.termux/files/usr/',
  GIT_EXEC_PATH: '/data/data/com.termux/files/usr/libexec/git-core/' }
```

It seems `resolveGitBinary()` always fails on an unsupported platform.

**Comment by shiftkey** [comment]
Looks like we need to do a bit more in here:

https://github.com/desktop/dugite/blob/73d4b3a134697445135b831204dc323398f360e9/lib/git-environment.ts#L27-L36

I gather we can map this to `{dir}/bin/git`?

**Comment by Rudloff** [comment]
This works on Android:
```diff
diff --git a/lib/git-environment.ts b/lib/git-environment.ts
index 5a97ce0..efb00cf 100644
--- a/lib/git-environment.ts
+++ b/lib/git-environment.ts
@@ -26,7 +26,7 @@ function resolveGitDir(): string {
  */
 function resolveGitBinary(): string {
   const gitDir = resolveGitDir()
-  if (process.platform === 'darwin' || process.platform === 'linux') {
+  if (process.platform === 'darwin' || process.platform === 'linux' || process.platform === 'android') {
     return path.join(gitDir, 'bin', 'git')
   } else if (process.platform === 'win32') {
     return path.join(gitDir, 'cmd', 'git.exe')
@@ -46,7 +46,7 @@ function resolveGitExecPath(): string {
     return path.resolve(process.env.GIT_EXEC_PATH)
   }
   const gitDir = resolveGitDir()
-  if (process.platform === 'darwin' || process.platform === 'linux') {
+  if (process.platform === 'darwin' || process.platform === 'linux' || process.platform === 'android') {
     return path.join(gitDir, 'libexec', 'git-core')
   } else if (process.platform === 'win32') {
     return path.join(gitDir, 'mingw64', 'libexec', 'git-core')
```

I'm not sure about other unsupported platforms though.

**Comment by shiftkey** [comment]
@Rudloff I'd accept that patch as a PR 😄 

> I'm not sure about other unsupported platforms though.

We can cross that bridge later.

**Comment by shiftkey** [comment]
@Rudloff this is available in `1.51.0` on NPM
